multicluster/controllers/multicluster/label_identity_export_controller.go

```diff
@@ -0,0 +1,325 @@
+// ResourceImport object if needed.
+func (r *LabelIdentityExportReconciler) cleanupLabelIdentityResourceImports(ctx context.Context,
+	clusterID, deletedClusterLabelHash string) error {
+	deletedCluster := sets.NewString(clusterID)
+	if clusters, ok := r.labelsToClusters[deletedClusterLabelHash]; ok && clusters.Equal(deletedCluster) {
+		// The cluster where the label identity is being deleted was the only cluster that has
+		// the label identity. Hence, the label identity is no longer present in the ClusterSet.
+		if !r.handleLabelIdentityDelete(ctx, deletedClusterLabelHash, clusterID) {
+			return fmt.Errorf("failed to delete LabelIdentity kind of ResourceImport for label %v of cluster %s", deletedClusterLabelHash, clusterID)
+		}
+	}
+	return nil
+}
+
+// handleLabelIdentityDelete deletes the ResourceImport of a label identity
+// hash that no longer exists in the ClusterSet. Note that the ID of a label
+// identity hash is only released if the deletion of its corresponding
+// ResourceImport succeeded.
+func (r *LabelIdentityExportReconciler) handleLabelIdentityDelete(ctx context.Context,
+	deletedLabelHash, clusterID string) bool {
+	labelIdentityImport := &mcsv1alpha1.ResourceImport{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      deletedLabelHash,
+			Namespace: r.namespace,
+		},
+	}
+	if err := r.Client.Delete(ctx, labelIdentityImport, &client.DeleteOptions{}); err != nil && !apierrors.IsNotFound(err) {
+		klog.ErrorS(err, "Failed to delete LabelIdentity kind of ResourceImport for stale label", "label", deletedLabelHash)
+		return false
+	}
+	// Delete caches for the label and release the ID assigned for the label
+	id := r.labelsToID[deletedLabelHash]
+	r.allocator.release(id)
+	delete(r.labelsToID, deletedLabelHash)
+	delete(r.labelsToClusters, deletedLabelHash)
+	if clusterLabels, ok := r.clusterToLabels[clusterID]; ok {
+		clusterLabels.Delete(deletedLabelHash)
+	}
+	return true
+}
+
+// handleLabelIdentityAdd creates ResourceImport of a label identity that is added in
+// the ClusterSet. Note that the ID of a label identity is only allocated and stored
+// if the creation of its corresponding ResourceImport succeeded.
+func (r *LabelIdentityExportReconciler) handleLabelIdentityAdd(ctx context.Context,
+	labelHash, label, clusterID string) bool {
+	labelIdentityResImport := &mcsv1alpha1.ResourceImport{}
+	ResImpKey := types.NamespacedName{Name: labelHash, Namespace: r.namespace}
+	if err := r.Client.Get(ctx, ResImpKey, labelIdentityResImport); err == nil {
+		idAllocated := labelIdentityResImport.Spec.LabelIdentity.ID
+		if err := r.allocator.setAllocated(idAllocated); err == nil {
+			return true
+		} else if err := r.Client.Delete(ctx, labelIdentityResImport, &client.DeleteOptions{}); err != nil {
+			// ResourceImport with stale label identity to ID mapping must be deleted.
+			klog.ErrorS(err, "Failed to delete outdated LabelIdentity kind of ResourceImport for label", "label", label)
```

Comment:
Why here we just log err? I assume this should return false immediately if it's not NotFound error?

Reply:
Fixed